### ISSUE TITLE
Hashes with symbol keys pushed to an array in node attributes do not allow string key access

### ISSUE THREAD

**Issue by b-dean** [chef/chef#6519]
## Description

When I push a hash that has symbol keys to an array that is in node attributes, I can no longer access them with string keys later. This previously worked in Chef 12 and 13.

## Chef Version

* Chef 12 versions `>= 12.21.10`
* Chef 13 versions `>= 13.4.19`

I believe this was introduced by the changes in #6355 #6356 #6357 #6358

## Platform Version

CentOS 7.4.1708

## Replication Case

Create `example.rb`
```ruby
node.normal['a'] = []
node.normal['a'].push({b: 1})
puts Chef::VERSION
puts "normal json: #{node.normal.to_json}",
     "a: #{node['a']}",
     "a class: #{node['a'].class}", 
     "a normal class: #{node.normal['a'].class}",
     "index 0: #{node['a'][0]}", 
     "index 0 class: #{node['a'][0].class}", 
     "index 0 normal class: #{node.normal['a'][0].class}",
     "b: #{node['a'][0]['b']}", 
     "b class: #{node['a'][0]['b'].class}",
     '*'*80
```

and run it with docker containers (seemed like an easy way to have a bunch of chef versions):
```
docker create --name chef-12.21.4 chef/chef:12.21.4
docker create --name chef-12.21.10 chef/chef:12.21.10
docker create --name chef-13.3.42 chef/chef:13.3.42
docker create --name chef-13.4.19 chef/chef:13.4.19
echo '--------------------------'
docker run --rm --volumes-from chef-12.21.4 -v `pwd`/example.rb:/example.rb centos  /opt/chef/bin/chef-apply /example.rb
docker run --rm --volumes-from chef-12.21.10 -v `pwd`/example.rb:/example.rb centos  /opt/chef/bin/chef-apply /example.rb
docker run --rm --volumes-from chef-13.3.42 -v `pwd`/example.rb:/example.rb centos  /opt/chef/bin/chef-apply /example.rb
docker run --rm --volumes-from chef-13.4.19 -v `pwd`/example.rb:/example.rb centos  /opt/chef/bin/chef-apply /example.rb
docker rm -v chef-12.21.4 chef-12.21.10 chef-13.3.42 chef-13.4.19
```

## Client Output

```
964bc889bf86a5196cd091e8b9349022caad3ecec7504d44ae38ca21743b2447
97e0dd9a45529c0e6a896f17edd5b24bce46d86c568fef508a39f18d86b7cb4f
8767dde5447f5a78e7afba7d03df49644a7ff5165291fe6430ab077f30aeed31
ba65e75600ff8c3fc475069ab961555a3a33f503b7f833223d3ba641eb889a33
--------------------------
[2017-10-24T15:18:56+00:00] INFO: The plugin path /etc/chef/ohai/plugins does not exist. Skipping...
[2017-10-24T15:18:56+00:00] WARN: Plugin Network: unable to detect ipaddress
[2017-10-24T15:18:57+00:00] INFO: Run List is []
[2017-10-24T15:18:57+00:00] INFO: Run List expands to []
12.21.4
normal json: {"tags":[],"a":[{"b":1}]}
a: [{"b"=>1}]
a class: Chef::Node::ImmutableArray
a normal class: Chef::Node::AttrArray
index 0: {"b"=>1}
index 0 class: Chef::Node::ImmutableMash
index 0 normal class: Hash
b: 1
b class: Fixnum
********************************************************************************
[2017-10-24T15:18:58+00:00] INFO: The plugin path /etc/chef/ohai/plugins does not exist. Skipping...
[2017-10-24T15:18:58+00:00] WARN: Plugin Network: unable to detect ipaddress
[2017-10-24T15:18:59+00:00] INFO: Run List is []
[2017-10-24T15:18:59+00:00] INFO: Run List expands to []
12.21.10
normal json: {"tags":[],"a":[{"b":1}]}
a: [{:b=>1}]
a class: Chef::Node::ImmutableArray
a normal class: Chef::Node::AttrArray
index 0: {:b=>1}
index 0 class: Chef::Node::ImmutableMash
index 0 normal class: Hash
b: 
b class: NilClass
********************************************************************************
[2017-10-24T15:19:00+00:00] WARN: Plugin Network: unable to detect ipaddress
13.3.42
normal json: {"tags":[],"a":[{"b":1}]}
a: [{"b"=>1}]
a class: Chef::Node::ImmutableArray
a normal class: Chef::Node::AttrArray
index 0: {"b"=>1}
index 0 class: Chef::Node::ImmutableMash
index 0 normal class: Hash
b: 1
b class: Integer
********************************************************************************
[2017-10-24T15:19:02+00:00] WARN: Plugin Network: unable to detect ipaddress
13.4.19
normal json: {"tags":[],"a":[{"b":1}]}
a: [{:b=>1}]
a class: Chef::Node::ImmutableArray
a normal class: Chef::Node::AttrArray
index 0: {:b=>1}
index 0 class: Chef::Node::ImmutableMash
index 0 normal class: Hash
b: 
b class: NilClass
********************************************************************************
chef-12.21.4
chef-12.21.10
chef-13.3.42
chef-13.4.19
```

**Comment by lamont-granquist** [comment]
can you confirm this happens on normal attributes and not default or override?

**Comment by b-dean** [comment]
I'm pretty sure it happens with all three, `default`, `normal`, `override`. I noticed it in our real code in something that was default level. For simplicity I only used `normal` in the example.

Here's something closer to what we actually are doing (in `chef-shell`):

```
chef (12.21.20)> node.default_unless['a']['b'] = []
 => [] 
chef (12.21.20)> node.default['a']['b'].push({c:1})
 => [{:c=>1}] 
chef (12.21.20)> node['a']['b'][0]['c']
 => nil 
chef (12.21.20)> node['a']['b'][0]
 => {:c=>1} 
```

that's with default. Interestingly, if I drop the second key out of there it seems to work: 

```
chef (12.21.20)> node.default_unless['a'] = []
 => [] 
chef (12.21.20)> node.default['a'].push({b:1})
 => [{:b=>1}] 
chef (12.21.20)> node['a'][0]
 => {"b"=>1} 
```

Same thing with `normal` and `override`

**Comment by lamont-granquist** [comment]
```
chef (13.6.1)> node.default['a'] = []
 => []
chef (13.6.1)> node.default['a'].push({b: "c"})
 => [{:b=>"c"}]
chef (13.6.1)> node.default['a']
 => [{:b=>"c"}]
chef (13.6.1)> node['a']
 => [{"b"=>"c"}]
```

okay so it looks like its getting stored that way at all three levels, but the deep-merge behavior of default and override levels is converting it (which i thought i made it not do, which is mildly confusing).

**Comment by lamont-granquist** [comment]
```
chef (13.6.1)> node.default['a'].class
 => Chef::Node::AttrArray
chef (13.6.1)> node.default['a'][0].class
 => Hash
```

ah.

**Comment by lamont-granquist** [comment]
the array seems to be important:

```
chef (13.6.4)> node.default['foo'] = { bar: "baz" }
 => {:bar=>"baz"}
chef (13.6.4)> node.default['foo']
 => {"bar"=>"baz"}
chef (13.6.4)> node.default['foo'].merge({ baz: "quux" })
 => {"bar"=>"baz", "baz"=>"quux"}
chef (13.6.4)> node.default['foo']
 => {"bar"=>"baz"}
chef (13.6.4)> node.default['foo'].merge!({ baz: "quux" })
 => {"bar"=>"baz", "baz"=>"quux"}
chef (13.6.4)> node.default['foo']
 => {"bar"=>"baz", "baz"=>"quux"}
chef (13.6.4)>
```

kind of odd that the return value of `#[]=` isn't converting the symbol, but the internal representation is changed.

so this seems to affect AttrArray's and not VividMashes.

```
chef (13.6.4)> node.default['foo'] = [ { bar: "baz" } ]
 => [{:bar=>"baz"}]
chef (13.6.4)> node.default['foo']
 => [{"bar"=>"baz"}]
chef (13.6.4)>
```

looks like `#[]=` on AttrArray is fine though.

**Comment by b-dean** [comment]
Maybe something in the [mutators](https://github.com/chef/chef/blob/master/lib/chef/node/attribute_collections.rb#L63-L69) should be calling `convert_value`?

**Comment by lamont-granquist** [comment]
Seems to be busted on 13.0.0 as well:

```
chef (13.0.1)> node.default[:foo] = []
 => []
chef (13.0.1)> node.default[:foo].push({bar: "baz"})
 => [{:bar=>"baz"}]
chef (13.0.1)> node.default[:foo]
 => [{:bar=>"baz"}]
chef (13.0.1)>
```

Still a bug, but it appears to not be a recent regression at any rate.

That makes some sense because I don't see anything in the mutators which are calling convert_value.  But I find it odd that it ever worked at all.

**Comment by lamont-granquist** [comment]
```
Ohai2u lamont@coredump.scriptkiddie.lan!
chef (12.5.1)> node.default[:foo] = []
 => []
chef (12.5.1)> node.default[:foo].push({bar: "baz"})
 => [{:bar=>"baz"}]
chef (12.5.1)> node.default[:foo]
 => [{:bar=>"baz"}]
chef (12.5.1)>
```

**Comment by lamont-granquist** [comment]
In 12.5.1 it looks like we had a consistently broken API:

```
chef (12.5.1)> node.default[:foo] = []
 => []
chef (12.5.1)> node.default[:foo].push({bar: "baz"})
 => [{:bar=>"baz"}]
chef (12.5.1)> node.default[:foo]
 => [{:bar=>"baz"}]
chef (12.5.1)> node[:foo]
 => [{:bar=>"baz"}]
chef (12.5.1)> node[:foo][0]["bar"]
 => nil
chef (12.5.1)> node[:foo][0][:bar]
 => "baz"
```

In 13.6.4 it looks like we now have an inconsistently fixed API:

```
chef (13.6.4)> node.default[:foo] = []
 => []
chef (13.6.4)> node.default[:foo].push({bar: "baz"})
 => [{:bar=>"baz"}]
chef (13.6.4)>  node.default[:foo]
 => [{:bar=>"baz"}]
chef (13.6.4)> node[:foo]
 => [{"bar"=>"baz"}]
chef (13.6.4)> node[:foo][0]["bar"]
 => "baz"
chef (13.6.4)> node[:foo][0][:bar]
 => "baz"
```

Only for the merged default and override attributes, we correctly convert symbols to strings now in the merged view.  Since normal attributes don't have the complex inter-precedence level array merging behavior they are still broken if you set those and try to retrieve them.

Since we went from only being able to access the keys as the type they were inserted as, to being able to access them differently this is mostly going to be non-breaking.

There's edge cases where this might have broken someone due to writes to `"foo"` and `:foo` *both* having been different and retrieved different values before, and now they overwrite each other.  Or if someone pulls the hash out of the array as a hash (#to_hash or whatever) it would have been rendered using symbols as keys and now it will be strings.

**Comment by b-dean** [comment]
That edge case of using both `"foo"` and `:foo` is exactly how we ran across it. Our cookbook's attributes files were adding hashes in arrays using the ruby 1.9 symbol key syntax (because it's arguably nicer looking) and then reading them later as string keys (because in our recipes we read all the keys as strings). Given that everything was supposed to be Mashes, this wasn't a problem before.

Should we maybe have never used symbols? I dunno, I guess. We've been using chef since the 0.10.x days and I don't know that there was a lot of consensus on the best way to do things back then. Also, in general in the Ruby world, it seems that people prefer symbols for keys so I don't know that we did anything particularly crazy. ¯\\\_(ツ)\_/¯

**Comment by lamont-granquist** [comment]
Yeah, [FC01](http://www.foodcritic.io/#FC001) suggests strings over symbols, and this is one reason why, since the internal representation is strings.

Dunno what to do about this because we fixed bugs and in the process we got this (somewhat tangentially related) behavior more instead of less correct.

Since we're getting reasonably close to Chef 14, I think this behavior gets addressed along with #6575, although i could argue these are bugs that need fixing.
